crates/block-producer/src/batch_builder/mod.rs

```diff
@@ -67,280 +66,246 @@ impl BatchBuilder {
-/// instead of returning None.
-struct WorkerPool {
-    in_progress: JoinSet<BatchResult>,
-    simulated_proof_time: Range<Duration>,
-    failure_rate: f32,
-    /// Maximum number of workers allowed.
-    capacity: NonZeroUsize,
-    /// Maps spawned tasks to their job ID.
-    ///
-    /// This allows us to map panic'd tasks to the job ID. Uses [Vec] because the task ID does not
-    /// implement [Ord]. Given that the expected capacity is relatively low, this has no real
-    /// impact beyond ergonomics.
-    task_map: Vec<(tokio::task::Id, BatchId)>,
-    store: StoreClient,
-    batch_prover: BatchProver,
-}
+        let job = BatchJob {
+            failure_rate: self.failure_rate,
+            store: self.store.clone(),
+            mempool,
+            batch_prover: self.batch_prover.clone(),
+        };
 
-impl WorkerPool {
-    fn new(
-        capacity: NonZeroUsize,
-        simulated_proof_time: Range<Duration>,
-        failure_rate: f32,
-        store: StoreClient,
-        batch_prover: BatchProver,
-    ) -> Self {
-        Self {
-            simulated_proof_time,
-            failure_rate,
-            capacity,
-            store,
-            in_progress: JoinSet::default(),
-            task_map: Vec::default(),
-            batch_prover,
-        }
+        self.worker_pool
+            .spawn(async move { job.build_batch().await }.instrument(tracing::Span::current()));
     }
 
-    /// Returns the next batch proof result.
-    ///
-    /// Will return pending if there are no jobs in progress (unlike tokio's [`JoinSet::join_next`]
-    /// which returns an option).
-    async fn join_next(&mut self) -> BatchResult {
-        if self.in_progress.is_empty() {
-            return std::future::pending().await;
+    #[instrument(target = COMPONENT, name = "batch_builder.wait_for_available_worker", skip_all)]
+    async fn wait_for_available_worker(&mut self) {
+        if let Err(crash) = self.worker_pool.join_next().await.expect("worker pool is never empty")
+        {
+            tracing::error!(message=%crash, "Batch worker pool panic'd");
+            panic!("Batch worker pool panic: {crash}");
         }
+    }
+}
 
-        let result = self
-            .in_progress
-            .join_next()
-            .await
-            .expect("JoinSet::join_next must be Some as the set is not empty")
-            .map_err(|join_err| {
-                // Map task ID to job ID as otherwise the caller can't tell which batch failed.
-                //
-                // Note that the mapping cleanup happens lower down.
-                let batch_id = self
-                    .task_map
-                    .iter()
-                    .find(|(task_id, _)| &join_err.id() == task_id)
-                    .expect("Task ID should be in the task map")
-                    .1;
-
-                (batch_id, join_err.into())
-            })
-            .and_then(|x| x);
+struct BatchJob {
```

Comment:
nit: let's add a section separator above this line.

Also, I would probably add some doc comments explaining the purpose/lifecycle of a `BatchJob`.